tests/test.py

```diff
@@ -475,8 +502,131 @@ def recur(j1, j2, path, unordered=False):
             )
 
 
-class JSONTestCase(WithJSONTestCase, TestCase):
-    """TestCase subclass with JSON comparisons and test enable/disable flag
-    handling."""
+class TruTestCase(WithJSONTestCase, TestCase):
+    """TestCase subclass with several additions.
+
+    - JSON comparisons and golden-file handling.
+
+    - Dumps of remaining tasks on completion of each test. This can optionally
+      result in test failures if `TEST_TASKS_CLEANUP` is set.
+
+    - Dumps non-main running threads after test completion. This can be made to
+      result in test failure if `TEST_THREADS_CLEANUP` is set.
+
+    - Garbage collection subtest. This test optionally prints out the reference
+      path to the given object that was not garbage collected. This is enabled
+      with the `WITH_REF_PATH` environment variable.
+    """
+
+    def assertCollected(self, ref: weakref.ReferenceType[T], msg=None):
+        """Check that the object referenced by `ref` has been garbage
+        collected.
+
+        Optionally prints out the reference path to the object if it was not
+        garbage collected. This is enabled with the `WITH_REF_PATH` environment
+        variable.
+        """
+
+        gc.collect()
+
+        if msg is None:
+            msg = f"Object {ref} was not garbage collected."
+
+        obj = ref()
+
+        with self.subTest(part="garbage collection"):
+            self.assertTrue(obj is None, msg)
+
+        # Enable WITH_REF_PATH to see printout of why the given ref was not
+        # GC-ed.
+        if (
+            obj is not None
+            and os.environ.get(WITH_REF_PATH_VAR, None) is not None
+        ):
+            caller_globals = python_utils.caller_frame(offset=1).f_globals
+
+            with self.subTest(part="reference path"):
+                # Show the reference path to the given ref.
+                print(f"Reference path from globals to {ref}:")
+                path = find_path(id(caller_globals), id(obj))
+                self.assertIsNotNone(path, "Couldn't find reference path.")
+                print_referent_lens(path)
+
+    def tearDown(self):
```

Comment:
might want to call the super version too since it may do some good stuff? I have no idea.

Reply:
added at bottom